mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java

```diff
@@ -0,0 +1,525 @@
+      while (true) {
+        if (currentIterator.hasNext()) {
+          currentRow = currentIterator.next();
+          return true;
+        } else if (tasks.hasNext()) {
+          currentCloseable.close();
+          currentIterator = open(tasks.next());
+        } else {
+          return false;
+        }
+      }
+    }
+
+    @Override
+    public Void getCurrentKey() {
+      return null;
+    }
+
+    @Override
+    public T getCurrentValue() {
+      return currentRow;
+    }
+
+    @Override
+    public float getProgress() {
+      return context.getProgress();
+    }
+
+    @Override
+    public void close() throws IOException {
+      currentCloseable.close();
+    }
+
+    private Iterator<T> open(FileScanTask currentTask) {
+      DataFile file = currentTask.file();
+      // schema of rows returned by readers
+      PartitionSpec spec = currentTask.spec();
+      Set<Integer> idColumns = spec.identitySourceIds();
+      Schema readSchema = expectedSchema != null ? expectedSchema : tableSchema;
+      boolean hasJoinedPartitionColumns = !idColumns.isEmpty();
+      if (hasJoinedPartitionColumns) {
+        readSchema = TypeUtil.selectNot(tableSchema, idColumns);
+        Schema identityPartitionSchema = TypeUtil.select(tableSchema, idColumns);
+        return Iterators.transform(
+            open(currentTask, readSchema),
+            row -> withPartitionColumns(row, identityPartitionSchema, spec, file.partition()));
+      } else {
+        return open(currentTask, readSchema);
+      }
+    }
+
+    private Iterator<T> open(FileScanTask currentTask, Schema readSchema) {
+      DataFile file = currentTask.file();
+      // TODO should we somehow make use of FileIO to create inputFile?
+      InputFile inputFile = HadoopInputFile.fromLocation(file.path(), context.getConfiguration());
+      CloseableIterable<T> iterable;
+      switch (file.format()) {
+        case AVRO:
+          iterable = newAvroIterable(inputFile, currentTask, readSchema);
+          break;
+        case ORC:
+          iterable = newOrcIterable(inputFile, currentTask, readSchema);
+          break;
+        case PARQUET:
+          iterable = newParquetIterable(inputFile, currentTask, readSchema);
+          break;
+        default:
+          throw new UnsupportedOperationException(
+              String.format("Cannot read %s file: %s", file.format().name(), file.path()));
+      }
+      currentCloseable = iterable;
+      if (inMemoryDataModel == InMemoryDataModel.DEFAULT) {
+        Expression filter = currentTask.residual();
+        return applyResidualsOnGenericRecords(iterable, filter).iterator();
+      } else {
+        return iterable.iterator();
+      }
+    }
+
+    private Iterable<T> applyResidualsOnGenericRecords(CloseableIterable<T> iterable, Expression filter) {
+      if (filter == null || filter.equals(Expressions.alwaysTrue())) {
+        return iterable;
+      } else {
+        throw new UnsupportedOperationException(
```

Comment:
Is it possible to throw this in `getSplits` so it isn't an error that may happen 95% of the way through a job?

Reply:
Fixed